src/backend/parser/parse_expr.c

```diff
@@ -2325,6 +2327,36 @@ transformCoalesceExpr(ParseState *pstate, CoalesceExpr *c)
 	{
 		Node	   *e = (Node *) lfirst(args);
 		Node	   *newe;
+		Oid		   etype = exprType(e);
+
+		/*
+		 *	T-SQL treats constant string literals as VARCHAR. Hence,
+		 *	coercing into VARCHAR before coercing it to the common type.
+		 */
+		if (sql_dialect == SQL_DIALECT_TSQL && !c->tsql_is_null && etype == UNKNOWNOID && IsA(e, Const))
+		{
+			Const	   *con = (Const *) e;
+			char	   *val = DatumGetCString(con->constvalue);
+			int	   i = -1;
+
+			if (val != NULL)
+				i = strlen(val) - 1;
+
+			/*
+			 *	Additional handling for empty or white space string literals as
+			 *	T-SQL treats an empty string literal as 0 in certain datatypes
+			 */
+			for (; i >= 0; i--)
+			{
+				if (val[i] != ' ')
+					break;
+			}
+
+			if (i != -1)
+				e = coerce_to_common_type(pstate, e,
+									 		VARCHAROID,
+									 		"COALESCE");
+		}
```

Comment:
Instead of adding this logic in the engine code, we should define a function in our babelfish extensions which contains this logic and just call that function.